src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/ILGen.cs

```diff
@@ -96,52 +96,46 @@ internal static void EmitLoadValueIndirect(this ILGenerator il, Type type)
         {
             ContractUtils.RequiresNotNull(type, nameof(type));
 
-            if (type.GetTypeInfo().IsValueType)
+            switch (type.GetTypeCode())
```

Comment:
I think at ILGen level using underlying type of enums is perfectly fine. An unboxed enum value is undistinguashable form its underlying type.
The codegen peculiarity is likely historical.
Also could be avoidance of fetching underlying enum type, if that is expensive, but it is doubtful.